### ISSUE TITLE
Lightkurve2 API - make LightCurve API less surprising

### ISSUE THREAD

**Issue by orionlee** [KeplerGO/lightkurve#845]
I did a small exercise, comparing the codes using Lightkurve2 `LightCurve` vs plain astropy `TimeSeries` in doing some lightcurve processing / plotting:
  https://gist.github.com/orionlee/633a42d34a8e0b6f166a85e35afa27d2

#### Observations
In current Lightkurve2 alpha, `LightCurve` object provides a set of shortcuts that 
1. expose underlying column data (e.g., `time`, `flux`, `quality` etc.) as attributes: e.g., `lc.quality`
2. expose `meta` items (e.g., `TSTART`, `SECTOR` in the underlying FITS header) as attributes: e.g., `lc.tstart`.

While they are quite convenient, it could be confusing when one wants to attach additional data to the object, e.g., detrended flux, relative time, etc. E.g., if one wants to add detrended flux to a LightCurve object by using `lc.flux_detrended`, it won't be treated as an additional column. 

Not treating as an column could lead to some behavior difference that could be surprising:
1. the data can't be used as a column in helpers such as `plot(column=...)`
2. When assigning data to a column, e.g., `lc['flux'] = flux_adjusted`, the values of `flux_adjusted` are copied to the `lc['flux']` column. Subsequent changes to `flux_adjusted` won't affect `lc['flux']` (and vice versa). In contrast, when data to an attribute, the data is not copied.

---

#### Suggested changes:
1. Eliminate attribute shortcuts that hide underlying meta or columns. Or provide read-only access, rather than read-write. 
  - astropy `TimeSeries` provides no such shortcut, with the lone exception of a read-only `.time`
  - when one tries to attach some data to LightCurve object, it will be clear what the code does
     - add / update columns, e.g. `lc['flux_detrended'] = ...`
     - add / update metadata, e.g., `lc.meta['tstart'] = ...`
     - add / update attributes is still possible

2. If (1) is adopted, consider to drop the automatic lowercasing of metadata key, e.g., accessing `TSTART` will become `lc.meta['TSTART']`
  - astropy `TimeSeries` does not do such conversion either.
  - For people familiar with Lightkurve 1 or FITS, the conversion could be a magic that surprises them.
  - I presume the motivation for automatic lowercasing is to make accessing them via attributes more natural (e.g., `lc.tstart`). With adoption of (1), either eliminating such attributes or make them read-only, I think the automatic lowercasing is no longer needed.

**Comment by barentsen** [comment]
@orionlee Thanks so much for sharing your perspective on this.  It is very important to carefully question the design decisions in v2.0 before we release a formal version. I really appreciate this! 🙏 

For the reasons you explain above, I now 100% agree that we should not have read-write attributes.  An additional reason not to allow write-access to attributes is that `TimeSeries` operations do not preserve attribute values (with the exception of `meta`).

For example:
```python
>>> ts = TimeSeries(time=Time([1,2,3], format='mjd'))
>>> ts.hello = 'world'
>>> ts.copy().hello
AttributeError: 'TimeSeries' object has no attribute 'hello'
```

Now ... should we get rid of attributes altogether, or make them read-only?

At this time, I feel in favor of keeping them (as read-only) for three reasons:
* My fingers hate typing `['COLUMN']`.  In particular, my pinkey struggles with `[` and `]`.
* I think a lot of people's pinkeys dislike `[` and `]`, which I suspect is why `pandas` supports read-only attribute shortcuts, and even part of the reason why Python's `namedtuple` and `dataclass` features exist.
* Lightkurve v1.x provided lc.flux, lc.flux_err, etc, so removing it would significant hurt backwards compatibility.

Are those valid reasons?  I'm happy to have my mind changed!

In conclusion: I guess my proposal is to modify `LightCurve.__setattr__` to emit a clear warning whenever a user assigns a value to *any* attribute, similar to `pandas`. Makes sense?

**Comment by barentsen** [comment]
Regarding forcing lowercase in `meta`, I struggle with this contradiction:
* I agree we should use uppercase for header fields which are traditionally always uppercase.
* But ... I dislike having `lc.mission` map onto `lc.meta['MISSION']`.

I guess we could apply the logic that the contents of `lc.meta` should preserve case (i.e. be uppercase), but that attributes should always be lowercase (because of PEP8).

In summary, I guess I might be in favor of getting rid of the forced lowercasing of `lc.meta` entries.

**Comment by orionlee** [comment]
> In conclusion: I guess my proposal is to modify LightCurve.__setattr__ to emit a clear warning whenever a user assigns a value to any attribute, similar to pandas. Makes sense?

To be clear, does the proposal mean:
1. Lightkurve emits a warning assigns a value to any attribute, including those represent underlying column / meta item? Or,
2. warns user when trying to assign a value to new attributes, and disallow (or allow with no warning) assignment for those backed by a column / meta item?

------
Regarding expose column / meta as attributes, I do find it convenient for typical use cases. We might want to do have a policy to avoid collision for edge cases. 

```
    lc['foo'] = ...
    lc.meta['foo'] = ...
    # what does lc.foo mean then?
```

It's probably quite unlikely in practice. But one can never guess what users would do.

------

> An additional reason not to allow write-access to attributes is that TimeSeries operations do not preserve attribute values (with the exception of meta).

Yeah, that's even worse, and would be surprising.

Astropy 4.1 introduced a mechanism of allowing attributes be preserved.
  https://docs.astropy.org/en/latest/table/construct_table.html?highlight=table#subclassing-table
(The doc is at 4.2, but I first found it in a local 4.1 doc).
I don't think it's really helpful for the cases we discussed here (end users creating ad-hoc ish attributes). Plus it seems like what it really does it to store the attribute in meta. It's not too unlike what Lightkurve2 tries to do by exposing meta as attributes.


------

> In summary, I guess I am in favor of getting rid of the forced lowercasing of lc.meta entries.

I agree that keeping lc.meta keyword uppercase probably give users least surprise.

As for making their attribute counterpart lowercased: it's a tad inconvenient, but it's not uncommon in various programming practices for cases to be changed to conformed to the convention. E.g., 
- accessing CSS from javascript, e.g., CSS attribute `font-family` would be exposed in javascript as `anElement.style.fontFamily` (changed to camel case)

**Comment by barentsen** [comment]
I double-checked the behavior of `pandas.DataFrame` and `astropy.Table`.

### What does pandas do?

In my earlier comment, I was wrong when I stated that `pandas` only provides read-only attributes.  `pandas` does provide write-access!  The only thing that is not allowed is creating a new column via attribute access.  For example:

```python
>>> import pandas as pd
>>> df = pd.DataFrame({'a': [1, 2, 3]})
>>> df.a   # read-access via attribute is allowed
0    1
1    2
2    3
Name: a, dtype: int64
>>> df.a = [4, 5, 6]   # write-access via attribute is allowed
>>> df
   a
0  4
1  5
2  6
>>> df.b = [7, 8, 9]  # creating a *new* column via attribute is not allowed
<stdin>:1: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
```

### What does AstroPy do?

In contrast, `astropy` Table and TimeSeries do not provide attribute access, with the exception of the `TimeSeries.time` attribute: 

```python
>>> from astropy.timeseries import TimeSeries
>>> from astropy.time import Time
>>> ts = TimeSeries({'time': Time([1, 2, 3], format='mjd'), 'a': [3, 4, 5]})
>>> ts.time  # read-access is only available for time
<Time object: scale='utc' format='mjd' value=[1. 2. 3.]>
>>> ts.time = Time([4, 5, 6], format='mjd')  # write-access not allowed for time
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: can't set attribute
>>> ts.a  # read-access not allowed for other columns
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'TimeSeries' object has no attribute 'a'
>>> ts.a = [7, 8, 9]   # this creates a property unrelated to the column
>>> ts
<TimeSeries length=3>
 time    a
object int64
------ -----
   1.0     3
   2.0     4
   3.0     5
```


### What is my proposal?

My preference would be to replicate the behavior of `pandas`.  It is way more user-friendly!  In case of a name collision between a column and a key of `lc.meta`, I propose the column gets the priority.

Like pandas, we should not allow creating a *new* column or `meta` entry via attribute to avoid confusion.  A clear warning message should be displayed in that case, which is what pandas does.

Does that sound reasonable?

(A dream is to eventually add attribute access to AstroPy Table objects to make them more similar to pandas.   This must have been discussed in the past, but I can't readily find a thread on the topic on AstroPy's GitHub.)

**Comment by orionlee** [comment]
The proposal sounds reasonable. The API doc would need an attribute access section to describe the behavior. 

I'd only augment it for the case  a name collision between a column and a key of `lc.meta`, we also emit a warning message, saying it is exposing the column rather than `lc.meta`.

**Comment by orionlee** [comment]
@barentsen I could work on a PR for the change the next week, if we agree to the API change: panda-like attribute access behavior, and retain the (upper) case of the keys in `meta`.

**Comment by barentsen** [comment]
@orionlee I have not started working on this -- please do if you're interested!

(PS: I haven't been able to give Lightkurve much attention this week, but aim to catch up with issues and PRs next week.  Thanks for your patience!)

**Comment by barentsen** [comment]
> if we agree to the API change: panda-like attribute access behavior, and retain the (upper) case of the keys in meta.

Yep, I continue to think this is the best solution!

**Comment by orionlee** [comment]
I'm checking to:
1. in case there are some early adopters of V2 that we should consult with before the change is made
2. to see if a PR can be reviewed and merged quickly if I have a PR next Monday / Tuesday. I'd rather have it quickly reviewed / merged, mainly because retaining the case of meta would touch involve changing many lines. I'd rather not have the PR hanging around for too long, as it'd then likely result in conflicts that need to be resolved.

**Comment by barentsen** [comment]
@orionlee I'd be able to give your PR full attention early next week! :+1:

We haven't made any full releases of 2.0 yet, so I think it's fair to make changes to the API.  A few beta versions have been pushed to pypi and can continue to be used should anyone rely on them.
